File: backend/routers/activities.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import datetime
from database import get_db
from models.activity import Activity
from models.activity_stream import ActivityStream

router = APIRouter()

@router.get("/")
async def list_activities(
    type: Optional[str] = None,
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    distance_bucket: Optional[str] = None,
    effort_zone: Optional[str] = None,
    terrain_type: Optional[str] = None,
    route_id: Optional[int] = None,
    sort_by: str = "start_date",
    sort_order: str = "desc",
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """List activities with filtering and pagination"""
    query = db.query(Activity)
    
    # Apply filters
    if type:
        query = query.filter(Activity.type == type)
    if start_date:
        query = query.filter(Activity.start_date >= start_date)
    if end_date:
        query = query.filter(Activity.start_date <= end_date)
    if distance_bucket:
        query = query.filter(Activity.distance_bucket == distance_bucket)
    if effort_zone:
        query = query.filter(Activity.effort_zone == effort_zone)
    if terrain_type:
        query = query.filter(Activity.terrain_type == terrain_type)
    if route_id:
        query = query.filter(Activity.route_id == route_id)
    
    # Apply sorting
    sort_column = getattr(Activity, sort_by, Activity.start_date)
    if sort_order == "desc":
        query = query.order_by(sort_column.desc())
    else:
        query = query.order_by(sort_column.asc())
    
    # Pagination
    total = query.count()
    activities = query.offset((page - 1) * per_page).limit(per_page).all()
    
    return {
        "activities": [
            {
                "id": a.id,
                "strava_id": a.strava_id,
                "name": a.name,
                "type": a.type,
                "sport_type": a.sport_type,
                "start_date": a.start_date.isoformat(),
                "moving_time": a.moving_time,
                "distance": a.distance,
                "total_elevation_gain": a.total_elevation_gain,
                "average_speed": a.average_speed,
                "average_heartrate": a.average_heartrate,
                "max_heartrate": a.max_heartrate,
                "average_cadence": a.average_cadence,
                "average_watts": a.average_watts,
                "suffer_score": a.suffer_score,
                "device_name": a.device_name,
                "distance_bucket": a.distance_bucket,
                "effort_zone": a.effort_zone,
                "terrain_type": a.terrain_type,
                "route_id": a.route_id,
                "has_streams": a.has_streams
            }
            for a in activities
        ],
        "pagination": {
            "page": page,
            "per_page": per_page,
            "total": total,
            "total_pages": (total + per_page - 1) // per_page
        }
    }

@router.get("/{activity_id}")
async def get_activity(activity_id: int, db: Session = Depends(get_db)):
    """Get activity details"""
    activity = db.query(Activity).filter(Activity.id == activity_id).first()
    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")
    
    return {
        "id": activity.id,
        "strava_id": activity.strava_id,
        "name": activity.name,
        "type": activity.type,
        "sport_type": activity.sport_type,
        "start_date": activity.start_date.isoformat(),
        "moving_time": activity.moving_time,
        "elapsed_time": activity.elapsed_time,
        "distance": activity.distance,
        "total_elevation_gain": activity.total_elevation_gain,
        "average_speed": activity.average_speed,
        "max_speed": activity.max_speed,
        "average_heartrate": activity.average_heartrate,
        "max_heartrate": activity.max_heartrate,
        "has_heartrate": activity.has_heartrate,
        "average_watts": activity.average_watts,
        "weighted_average_watts": activity.weighted_average_watts,
        "max_watts": activity.max_watts,
        "average_cadence": activity.average_cadence,
        "suffer_score": activity.suffer_score,
        "kilojoules": activity.kilojoules,
        "device_name": activity.device_name,
        "gear_id": activity.gear_id,
        "distance_bucket": activity.distance_bucket,
        "effort_zone": activity.effort_zone,
        "terrain_type": activity.terrain_type,
        "route_id": activity.route_id,
        "created_at": activity.created_at.isoformat(),
        "updated_at": activity.updated_at.isoformat()
    }

@router.get("/{activity_id}/streams")
async def get_activity_streams(
    activity_id: int,
    stream_types: Optional[str] = None,
    resolution: str = "medium",
    db: Session = Depends(get_db)
):
    """Get activity streams (time-series data)"""
    activity = db.query(Activity).filter(Activity.id == activity_id).first()
    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")
    
    query = db.query(ActivityStream).filter(ActivityStream.activity_id == activity_id)
    
    if stream_types:
        types = stream_types.split(",")
        query = query.filter(ActivityStream.stream_type.in_(types))
    
    streams = query.all()
    
    return {
        "activity_id": activity_id,
        "streams": {
            s.stream_type: {
                "data": s.data,
                "series_type": s.series_type,
                "original_size": s.original_size,
                "resolution": s.resolution
            }
            for s in streams
        }
    }

File: frontend/src/api/activities.ts
import apiClient from './client'

export interface ActivitySummary {
  id: number
  strava_id: number | null
  name: string
  type: string
  sport_type: string | null
  start_date: string
  start_date_local: string
  moving_time: number | null
  distance: number | null
  total_elevation_gain: number | null
  average_speed: number | null
  average_heartrate: number | null
  max_heartrate: number | null
  average_cadence: number | null
  average_watts: number | null
  suffer_score: number | null
  device_name: string | null
  distance_bucket: string | null
  effort_zone: string | null
  terrain_type: string | null
  route_id: number | null
  has_streams: boolean
}

export interface ActivityDetail extends ActivitySummary {
  elapsed_time: number | null
  max_speed: number | null
  has_heartrate: boolean
  weighted_average_watts: number | null
  max_watts: number | null
  kilojoules: number | null
  gear_id: string | null
  computed_metrics: Array<{
    metric_type: string
    value: number
    computed_at: string
  }>
  effort_groups: Array<{
    group_type: string
    group_label: string
    group_value: string
    time_in_zone: number | null
  }>
  created_at: string
  updated_at: string | null
}

export interface Pagination {
  page: number
  per_page: number
  total: number
  total_pages: number
}

export interface ActivitiesResponse {
  activities: ActivitySummary[]
  pagination: Pagination
}

export interface ActivityFilters {
  type?: string
  start_date?: string
  end_date?: string
  distance_bucket?: string
  effort_zone?: string
  terrain_type?: string
  route_id?: number
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  page?: number
  per_page?: number
}

export interface EffortZoneBreakdown {
  zone: string
  label: string
  time_seconds: number
  percentage: number
}

export interface EffortData {
  activity_id: number
  max_hr: number
  dominant_zone: string
  total_time: number
  zone_breakdown: EffortZoneBreakdown[]
}

export interface CalendarDay {
  date: string
  value: number
  count: number
}

export interface CalendarSummary {
  total_activities: number
  total_distance: number
  total_moving_time: number
  longest_streak: number
  most_active_day: string
}

export interface CalendarResponse {
  metric: string
  data: CalendarDay[]
  start_date: string
  end_date: string
  summary: CalendarSummary
}

export const activitiesApi = {
  async list(filters: ActivityFilters = {}): Promise<ActivitiesResponse> {
    const params = new URLSearchParams()
    
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.set(key, String(value))
      }
    })
    
    const response = await apiClient.get<ActivitiesResponse>(
      `/activities/?${params.toString()}`
    )
    return response.data
  },

  async getCalendar(params: {
    start_date?: string
    end_date?: string
    metric?: 'distance' | 'count' | 'moving_time'
  } = {}): Promise<CalendarResponse> {
    const searchParams = new URLSearchParams()
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        searchParams.set(key, String(value))
      }
    })
    const response = await apiClient.get<CalendarResponse>(
      `/activities/calendar?${searchParams.toString()}`
    )
    return response.data
  },

  async getDetail(activityId: number): Promise<ActivityDetail> {
    const response = await apiClient.get<ActivityDetail>(
      `/activities/${activityId}`
    )
    return response.data
  },

  async delete(activityId: number): Promise<void> {
    await apiClient.delete(`/activities/${activityId}`)
  },

  async getEffort(activityId: number): Promise<EffortData> {
    const response = await apiClient.get<EffortData>(
      `/activities/${activityId}/effort`
    )
    return response.data
  },
}
